dev/bus/export/dev/serial.hpp

```diff
@@ -0,0 +1,217 @@
+    //!                     successfully sent before error occur.
+    //! \return Operation status.
+    static err send_buf(const uint8_t *buf, size_t &sz);
+
+private:
+    static void bus_handler(bus_channel ch, bus_event type, size_t total);
+
+    static err try_start_xfer();
+
+    static uint8_t m_rx_buffer[buf_size];
+    static uint8_t m_tx_buffer[buf_size];
+
+    static binary_semaphore m_rx_new_xfer_allowed;
+
+    static std::atomic_size_t m_rx_write_iter; //!< Used for accepting data from the bus
+    static std::atomic_size_t m_rx_read_iter; //!< Used for returning data to the user
+
+    static bool m_is_inited;
+};
+
+template <class PBus, size_t buf_size>
+uint8_t serial<PBus, buf_size>::m_rx_buffer[buf_size];
+
+template <class PBus, size_t buf_size>
+uint8_t serial<PBus, buf_size>::m_tx_buffer[buf_size];
+
+template <class PBus, size_t buf_size>
+std::atomic_size_t serial<PBus, buf_size>::m_rx_write_iter;
+
+template <class PBus, size_t buf_size>
+std::atomic_size_t serial<PBus, buf_size>::m_rx_read_iter;
+
+template <class PBus, size_t buf_size>
+bool serial<PBus, buf_size>::m_is_inited;
+
+template <class PBus, size_t buf_size>
+binary_semaphore serial<PBus, buf_size>::m_rx_new_xfer_allowed;
+
+template <class PBus, size_t buf_size>
+err serial<PBus, buf_size>::init()
+{
+    ecl_assert(m_is_inited == false);
+    auto result = PBus::init();
+    if (is_error(result)) {
+        return result;
+    }
+    PBus::set_handler(bus_handler);
+    result = PBus::enable_listen_mode();
+    if (is_error(result)) {
+        return result;
+    }
+    m_rx_new_xfer_allowed.signal();
+    PBus::set_rx(m_rx_buffer, buffer_size);
+    result = PBus::do_xfer();
+    if (is_ok(result)) {
+        m_is_inited = true;
+    }
+    return result;
+}
+
+template <class PBus, size_t buf_size>
+void serial<PBus, buf_size>::bus_handler(bus_channel ch,
+        bus_event type, size_t total)
+{
+    if (ch == bus_channel::rx) {
+        ecl_assert(type == bus_event::tc);
+        m_rx_write_iter++;
+        ecl_assert(m_rx_write_iter == total);
+
+        // Start new xfer if needed and allowed
+        if (m_rx_new_xfer_allowed.try_wait()) {
+            try_start_xfer();
+            // Allow user to read from rx_buffer
+            m_rx_new_xfer_allowed.signal();
+        }
+    }
+    // TODO: handle TX events
+}
+
+template <class PBus, size_t buf_size>
+err serial<PBus, buf_size>::recv_byte(uint8_t &byte)
+{
+    size_t sz = 1;
+    return recv_buf(&byte, sz);
+}
+
+template <class PBus, size_t buf_size>
+err serial<PBus, buf_size>::recv_buf(uint8_t *buf, size_t &sz)
+{
+    ecl_assert(m_is_inited == true);
+    // Restrict starting new xfer from bus_handler()
+    // to prevent data loss
+    m_rx_new_xfer_allowed.wait();
+
+    // Wait until data is available on the read end
+    while ((m_rx_read_iter == m_rx_write_iter)) { }
+
+    sz = std::min(sz, m_rx_write_iter - m_rx_read_iter);
```

Comment:
`m_rx_write_iter` can be modified during execution of this code. Copy it onto stack to avoid it